### ISSUE TITLE
Optionally preserve original hardlinks

### ISSUE THREAD

**Issue by gtristan** [ostreedev/ostree#925]
So first, I realize this may be a non-issue for ostree-as-host, where the design of ostree-as-host is (IIUC) to have a readonly filesystem managed by ostree with all hardlinks into the repository. However, fixing this can be very useful to abstract build systems which use ostree for caching of build results.

The problem I see here is that:
* OSTree at least at checkout time will hardlink every file which only differs in name (same attributes and content)
* This means that typically, all empty files with 0644 will be hardlinks to eachother in a checkout, notably this includes most `__init__.py` files in python packages
* When deploying a system created with all files linked in this way, it becomes possible to corrupt the system quite easily at boot time

An example of this scenario is where you deploy an OS which uses systemd, and systemd decides to write out a hex string to `/etc/machine-id`, and suddenly no python packages work anymore and the system doesnt completely boot.

Note that this issue is less severe seeing as we now have an option with OSTree to force copies at checkout time, meaning we can reduplicate any files which OSTree deduplicated, but this still means there is some lost information about which file was linked together at checkin time, which can no longer be retrieved at checkout time.

For instance it is quite plausible that a given system has an expectation that some files be hard linked, a system may desire to provide the same file at multiple locations and then intentionally write to that file.

This is relatively low priority for me at this time but I thought it would be better to keep this on the radar.

**Comment by cgwalters** [comment]
Yes, the `OstreeSysroot` level creates copies for files in `/etc`.  Why are you using bootable ostrees without the rest of the ostree-as-host infrastructure?

**Comment by cgwalters** [comment]
Is it something like wanting to reuse host ostree content as buildroots or for testing in containers?  In that case (particularly the testing one), https://mail.gnome.org/archives/ostree-list/2017-February/msg00007.html may be interesting.

But for the buildroot case, what gcontinuous does IIRC is symlink `etc` → `usr/etc` and basically has a read-only bind mount over everything except for `/tmp` and a host bind mount for the build result location.

**Comment by gtristan** [comment]
Not exactly no.

We are using ostree only for revisioning of filesystem data, as a storage backend for build outputs (generally the output data of `make DESTDIR=${install-root} install`). OSTree is very interesting to us simply as a sort of git for abstract filesystem data.

Hardlink checkouts make things optimized whenever we need to checkout a given revision of a build, and we generally save a lot of space this way compared with other techniques (like storing a tarball for each build output).

What gets built *might* also be deployed into an OSTree-as-host system or OSTree specific container, but our aim is *only* a build system. We need to keep this completely decoupled with any kind of deployment or packaging system, so the storage unit for build outputs should not make any assumptions about what kind of environment the given stored software will run in.

I realize that this is not the intent of what OSTree was originally designed for, but in every respect as far as I can see, the subset of OSTree functionality that is just storing revisioned files seems to be the best choice available.

**Comment by cgwalters** [comment]
That type of thing is definitely part of the original idea; it's really the reason why `OstreeRepo` and `OstreeSysroot` are distinct layers.

But anyways, let's dive into details here - you mentioned `/etc/machine-id`.  Does it work for you to simply mount `/etc` read-only via `bwrap`?  Or do you need to support mutation in place?  If so, have you tried the `rofiles-fuse` that comes with ostree?

**Comment by cgwalters** [comment]
Today for example [rpm-ostree uses rofiles-fuse](https://github.com/projectatomic/rpm-ostree/blob/4a61e556636f0b2be794611518db9d5266b1380b/src/libpriv/rpmostree-scripts.c#L201) for package layering scripts.

**Comment by gtristan** [comment]
Right, I do need mutation in place in a couple of places:
* When staging dependencies for a build, before the build we run things in the sysroot which can modify files (things like fc-cache, ldconfig, glib-compile-schemas, that kind of stuff)
* Once the above is done, when running a build we have the rootfs readonly and mount the build directory read-write for better performance
* We have other kinds of elements which might want to mutate things in other ways than building something, for instance we have this [script element](https://gitlab.com/BuildStream/gnome-modulesets-base/blob/master/elements/base-configure.bst) which will run `dpkg --configure` on an imported multistrap system (this lets us multistrap into a revisioned OSTree repo for multiple arches and just do the other steps on a host compatible arch at build time)

So, I was aware of rofiles-fuse, but I didnt want to have to care about special casing things in build scripts by explicitly making things fail. Instead I came up with this [fuse layer](https://gitlab.com/BuildStream/buildstream/blob/master/buildstream/_fuse/hardlinks.py) which the sandbox uses to provide a copy-on-write hardlink experience for the sandbox, so if ever an element needs to write something which was staged with hardlinks, it won't corrupt the artifact cache (ultimate an OSTree storage in bare-user mode)

The thing is basically, this is not a serious problem for the build sandbox, the copy-on-write hardlinks experience can be enhanced so that hardlink chains staged into the sysroot are maintained while separating those from the outside ostree store easily enough; but without knowing what those intended chains were at commit time this enhancement would be moot (every file with the same content and attributes is deduplicated into hardlinks so we cant know).

It's not in our domain to know what the user wanted to build ultimately, so when it comes time to create some output from a build pipeline, currently we would have to ensure that nothing is hardlinked to get the safest result (or work around this with extra, separately committed metadata); because we can't know what the user intended (did they want to build a system with a readonly `/usr` and symlinked `/etc -> /usr/etc` ? only maybe).

**Comment by cgwalters** [comment]
`rofiles-fuse` *does* support "in place" mutation, it just requires the create-new-then rename pattern.  Obviously not everything does that, but enough does that I've been working on pushing things over the hump.  You can find a few patches in the [rpm-ostree tracker](https://bugzilla.redhat.com/show_bug.cgi?id=1352154).

We could add copy-on-write to `rofiles-fuse`; some discussion on that in https://bugzilla.redhat.com/show_bug.cgi?id=1415451

I'm going to retitle this bug to that effect; I know this is a wide-ranging discussion but given you already have something similar, I think it'd probably make sense to add as an option.  We can break out anything else as separate issues?

**Comment by gtristan** [comment]
> I'm going to retitle this bug to that effect

I dont think that is appropriate, rather this support for copy-on-write hardlinks in rofiles-fuse is a separate topic that is only marginally related to the issue I'm trying to raise.

Can we please change this bug title back ? This is not a huge priority for me in the immediate future but I would like to keep a record of it on the radar if that is okay.

In an attempt to clarify: hardlinks which were present at commit time are impossible to retrieve at checkout time. It is not correct to say that in an abstract system either:
* Any file that is the same, can be hardlinked together, and no damage will be done
* Any file that was a hardlink, can be replaced with a copy and no damage will be done

But currently, as we cannot preserve the originally intended hardlinks at commit time, we cannot restore them at checkout time, so we are left with either of the choices above, regardless of what we intend to deploy.

In my first comment/report, I noted how the former case can be destructive in a system which does not use a read-only `/usr` and symlinked `/etc -> /usr/etc` setup (which is a traditional and still valid setup), but also on the other hand, there are cases where one file intentionally appears at multiple locations as a hardlink (for a random example, `dpkg --configure -a` happens to create a lot of hardlinks intentionally) and without making any assumptions about how the built system is supposed to run, it may even be that writing to a hardlinked file is intentional.

In a system that runs with ostree-as-host, perhaps you can assert one of the above, but we do not require that builds which happened to have used ostree internally for caching of build outputs, be builds of systems which run ostree-as-host, they can be builds of anything at all.

Below is another little script which demonstrates the issue I'm trying to get at:

```bash

#!/bin/bash

# Test case demonstrating that hardlink information which was present
# at commit time is lost unconditionally at checkout time.
#
# Usage:
#
#  ./test.sh <work directory> [<repomode>]
#
# Note that results differ if specified mode is 'bare' or 'bare-user'
# or 'archive-z2', but in every case, the input hardlink references
# are lost at checkout time.
#
# Note about test data:
#
#   a.txt - an empty regular file 0644
#   b.txt - another empty regular file 0644, that is a different file
#           from a.txt
#   c.txt - a hardlink to b.txt, b.txt and c.txt are the same file
#
set -e

topdir=${1}
repomode=${2}
if [ -z "${repomode}" ]; then
    repomode="bare-user"
fi

mkdir -p "${topdir}"
topdir="$(cd ${topdir} && pwd)"

repo="${topdir}/repo"
checkin="${topdir}/checkin"
checkout="${topdir}/checkout"
checkout_copy="${topdir}/checkout_copy"

function print_dir() {
    dir=$1
    description=$2

    echo "Printing ${description}: ${dir}"
    for file in $(find "${dir}" -type f | sort); do
	inode=$(stat -c "%i" "${file}")
	links=$(stat -c "%h" "${file}")
	echo "  $(basename ${file}) - inode: ${inode} links: ${links}"
    done
}

#
# Create test data with some permissions
#
mkdir -p "${checkin}"
touch "${checkin}/a.txt"
touch "${checkin}/b.txt"
chmod 0644 "${checkin}/a.txt"
chmod 0644 "${checkin}/b.txt"
ln "${checkin}/b.txt" "${checkin}/c.txt"

#
# Init repo with bare-user
#
ostree init --repo "${repo}" --mode "${repomode}"

#
# Create commit
#
ostree commit --repo "${repo}" --branch "test" -s "test commit" "${checkin}"

#
# Print the initial commit directory
#
print_dir "${checkin}" "initial commit"

#
# Checkout commit in user mode
#
ostree checkout --repo "${repo}" --user-mode --disable-cache "test" "${checkout}"
print_dir "${checkout}" "regular user checkout"

#
# Checkout commit with --force-copy
#
ostree checkout --repo "${repo}" --user-mode --disable-cache --force-copy "test" "${checkout_copy}"
print_dir "${checkout_copy}" "regular user --force-copy checkout"

echo
echo "~~~~~~~~~~~~"
echo "Now lets use tar to simulate what happens when we take these"
echo "checkouts away from ostree and deploy them in some abstract way,"
echo "including the original checkin"
echo "~~~~~~~~~~~~"
echo

function untar() {
    tarball=$1
    base="$(basename ${tarball})"
    base="${base%.tar}"
    unpack="${topdir}/${base%.tar}_unpack_tar"

    mkdir "${unpack}"
    cd "${unpack}"
    tar -xf "${tarball}"

    echo "${unpack}"
}

cd "${checkin}" && tar -cf "${topdir}/checkin.tar" .
cd "${checkout}" && tar -cf "${topdir}/checkout.tar" .
cd "${checkout_copy}" && tar -cf "${topdir}/checkout_copy.tar" .

dir=$(untar "${topdir}/checkin.tar")
print_dir "${dir}" "untar of initial commit"
echo ""
echo "  Note that the hardlink of b.txt & c.txt is preserved"
echo ""

dir=$(untar "${topdir}/checkout.tar")
print_dir "${dir}" "untar of regular user checkout"
echo ""
echo "  Note that if hardlinks were checked out, these are now separate"
echo "  from the ostree hardlinks, but a.txt, b.txt and c.txt are the"
echo "  same file, the committed hardlinks are lost"
echo ""

dir=$(untar "${topdir}/checkout_copy.tar")
print_dir "${dir}" "untar of regular user --force-copy checkout"
echo ""
echo "  In the case of the --force-copy checkouts, they seem to be always"
echo "  separate files, so we lose the original link between b.txt and c.txt"
echo ""
```

**Comment by cgwalters** [comment]
OK.  We could perhaps try to record the original hardlinks optionally in commit metadata and restore them.

I dunno though, couldn't you just use `ostree checkout --force-copy` unless you know it's ostree-as-host?  Actually are you even scoping in ostree-as-host into the buildsystem?  

I bet if we avoided hardlinking empty files it'd solve a lot of cases.

**Comment by gtristan** [comment]
Thanks for retitling this.

> I dunno though, couldn't you just use `ostree checkout --force-copy` unless you know it's ostree-as-host? 

Yes we can force everything in a checkout to become a copy, either by copying the checkout directory recursively or by using the `--force-copy` option, however this means that nobody can ever intentionally deploy file "foo" and "bar" as the same inode (one could then no longer ever intentionally deploy hardlinks).

> I bet if we avoided hardlinking empty files it'd solve a lot of cases.

If we avoided hardlinking empty files specifically, this assumes that two identical files which were committed as separate files, could never possibly be written to (perhaps they were two identical configuration templates).

> Actually are you even scoping in ostree-as-host into the buildsystem?

Quite intentionally not, one of our objectives is to ensure that the build system is decoupled from any specific kind of deployment at all.

With that said, ostree-as-host is something we very much want to use for the Baserock project, and there has been some work towards that already. Baserock being a separate project where we build a reference linux systems from scratch (also looking into cross bootstrapping to new hardwares with this project).

So to clarify, ostree-as-host is something that we want BuildStream to be battle tested for (via Baserock), definitely. But it should by no means have special cases and code within the build system to support (the fact that we use ostree as an artifact caching system internally should not impact our ability to roll out output into an ostree-as-host system, or other kinds of deployment).

## Thoughts on an plausible implementation

After some further thought, I think that a patch for this might simple enough, too.

Consider how `tar` works with hardlinks: An internal inode number (separate from the actual disk inode) is recorded for every file; when encountering multiple hardlinked files at archive creation time, the same internal inode is used. At extract time, the internal inode number is used to create hardlinks instead of copies.

For ostree this can be pretty simple; if I am starting to understand the underlying mechanics correctly, currently: *Everything with the same attributes and content is treated as the same file object, regardless of path and filename*

From what I understand, we would need three adjustments:
1. At commit time, every file records an internal inode number starting from 0, incrementing that internal number for every file being committed.
    * During the commit session, any files which had the same inode in the commit are assigned the same internal inode in the commit metadata instead of incrementing the internal inode number
2. Commits should traverse the filesystem in a deterministic order
3. Force copy checkout, I think should should observe the inodes also, and create hardlinks for files in the checkout destination to already checked out files with the same virtual inode number (this seems to me correct, what is checked out is force copied *out of the repo*, that does not necessarily mean intentional hardlink chains should be lost).

Since hardlinking is already entirely automated for everything that has the same metadata and content, just recording a virtual inode number (1) is enough to distinguish files which were hardlinks, and the hardlinked checkouts would automatically be correct.

We would want (2) purely as an optimization (but an important one). Ensuring that we iterate over the directory being committed in a determinstic order would mean that we always assign the same inode number for any subsequent commits in a branch, so we get to retain that information of whether or not a file has changed between two commits.

I think that after this:
* bare repo modes would be heaver on disk, because there would be less automatic deduplication (of course, the committer can always deduplucate to hardlinks themselves before committing, in the cases that they know this is safe, or ostree could always make this optional).
* archive repos theoretically should not suffer much impact in size, as added redundant data should in theory compress well.

**Comment by cgwalters** [comment]
The thing is that the per-file metadata is (somewhat by design) not extensible.  If it was extensible, then checksums would be different.

Storing additional data in the commit metadata would be simplest, but nontrivial, since `ostree-repo-checkout.c` would need to learn about it. Adding a new e.g. `OSTREE_OBJECT_TYPE_FILE_EXT` or something would be...a big endeavor.

Is it only hardlinks you're concerned about, or is it also mtime?

**Comment by gtristan** [comment]
I see, that is a bit unfortunate, it looks undesirable to add new object types and handle them in branches all over the codebase, instead it would be nicer to have the file object handle additional metadata. I am tempted to think that a custom/derived GFile/GFileInfo could be enough to carry any extra metadata through the existing OSTREE_OBJECT_TYPE_FILE objects, is that not feasible ? (I'm guessing that it's not possible for some reason).

Perhaps that would be a backwards incompatible approach (in terms of repo format) ?

Regarding mtimes... I was under the impression that mtime was revisioned but only skipped at checkout time (overridden with hard coded timestamp). Is mtime at commit time also irretrievable (at least by walking the metadata without doing a checkout) ?

Most of the time, we intentionally want a homogeneous mtime for build outputs, but we want to experiment with an idea later this year which would require preserving timestamps of build *trees* (the objects and sources together). That idea would be, instead of *only* caching the `make DESTDIR=${install-root} install` output of a module, we could additionally cache the build directory itself.

This would allow some interesting functionality, for instance:
* I found a bug on this generated operating system
* I know the cache keys for all the modules which went into this system
* Now let me just setup a workspace with the exact WebKit source and build tree which produced it's output for this cache key
* Modify one line of code
* Build only the parts of webkit which depend on the file which I modified (why we would consider caching a build tree, so here we would need preserved mtimes)
* Deploy the whole system as an image in the shortest possible time, to test this small change

**Comment by cgwalters** [comment]
We don't store mtimes at all; it's not part of the data format.

The `GFile{,Info}` bits are generated from the underlying GVariant formats, see `OSTREE_FILEMETA_GVARIANT_FORMAT` and `OSTREE_DIRMETA_GVARIANT_STRING`.  Neither are extensible  - again intentionally, because the ostree format (like git) is designed to be reproducibly checksummed, and extensibilty clashes with that.

**Comment by gtristan** [comment]
> Neither are extensible - again intentionally, because the ostree format (like git) is designed to be reproducibly checksummed, and extensibilty clashes with that.

That makes perfect sense yes, any approach is going to require some branchy code.
